packages/transaction-controller/src/gas-flows/SmartTransactionGasFeeFlow.ts

```diff
@@ -0,0 +1,132 @@
+import { hexToBN, toHex } from '@metamask/controller-utils';
+import { createModuleLogger, type Hex } from '@metamask/utils';
+import type BN from 'bn.js';
+
+import { projectLogger } from '../logger';
+import type {
+  GasFeeEstimates,
+  GasFeeFlow,
+  GasFeeFlowRequest,
+  GasFeeFlowResponse,
+  TransactionMeta,
+  getSmartTransactionFeeEstimatesResponse,
+} from '../types';
+import { GasFeeEstimateLevel } from '../types';
+import { pickMiddleFeeElement } from '../utils/utils';
+import { DefaultGasFeeFlow } from './DefaultGasFeeFlow';
+
+const log = createModuleLogger(projectLogger, 'smart-transaction-gas-fee-flow');
+
+const BASE_FEE_MULTIPLIERS = {
+  low: 1,
+  medium: 1.35,
+  high: 1.7,
+};
+
+const PRIORITY_FEE_MULTIPLIERS = {
+  low: 1,
+  medium: 1.05,
+  high: 1.1,
+};
+
+type FeesByLevel = {
+  [key in GasFeeEstimateLevel]: BN;
+};
+
+/**
+ * Implementation of a smart transaction gas fee flow specific to Smart Transactions.
+ */
+export class SmartTransactionGasFeeFlow implements GasFeeFlow {
+  matchesTransaction(transactionMeta: TransactionMeta): boolean {
+    return Boolean(transactionMeta.isSmartTransaction);
+  }
+
+  async getGasFees(request: GasFeeFlowRequest): Promise<GasFeeFlowResponse> {
+    try {
+      return await this.#getSmartTransactionGasFees(request);
+    } catch (error) {
+      log('Using default flow as fallback due to error', error);
+      return new DefaultGasFeeFlow().getGasFees(request);
+    }
+  }
+
+  async #getSmartTransactionGasFees(
+    request: GasFeeFlowRequest,
+  ): Promise<GasFeeFlowResponse> {
+    const smartTransactionFeesResponse = (await this.#getGasFeesAPIResponse(
+      request,
+    )) as NonNullable<getSmartTransactionFeeEstimatesResponse['tradeTxFees']>;
+
+    log('Received smart transaction response', smartTransactionFeesResponse);
+
+    const baseValue = pickMiddleFeeElement(smartTransactionFeesResponse.fees);
+
+    log('Picked base value', baseValue);
+
+    const maxFees = this.#getValuesFromMultipliers(
```

Comment:
Sorry for the confusion, but we shouldn't need any custom logic in this fee flow, rather we just want to return a single `medium` value by using the highest value of the smart transaction fees.

This will mean we can make `low` and `high` optional in `GasFeeEstimates`.

Then when we adopt in future in the extension, we can disable the custom button if there is no `low` and `high`.

Reply:
Done in https://github.com/MetaMask/core/pull/4009/commits/c57cfd91bc45e7ef7e02329d100193a21032fa86